torch/csrc/jit/passes/guard_elimination.cpp

```diff
@@ -122,36 +129,126 @@ struct GuardElimination {
     }
   }
 
- private:
-  bool removableGuard(Node* n) {
-    if (!simple_ops_.count(n->kind())) {
-      return false;
-    }
-
+  bool checkInputs(Node *n, const std::unordered_set<size_t> &except) {
     bool all_inputs_guarded = true;
+    size_t i = 0;
     for (auto input : n->inputs()) {
-      if (input->node()->kind() == prim::Guard ||
-          input->node()->kind() == prim::Constant) {
+      if ((input->node()->kind() == prim::Guard &&
+           !input->type()->expect<TensorType>()->isSummarized()) ||
+          input->node()->kind() == prim::Constant ||
+          input->type()->isSubtypeOf(NumberType::get()) ||
+          except.count(i) != 0) {
         AT_ASSERT(
             input->node()->kind() != prim::Guard ||
             input->type()->expect<TensorType>());
       } else {
+        GRAPH_DEBUG("input ", input->debugName(), " isn't guarded, type ",
+                    *input->type());
         all_inputs_guarded = false;
         break;
       }
+      i++;
     }
     return all_inputs_guarded;
```

Comment:
is this the right name for the function?

Reply:
`checkInputInvariants`? I think the current name is both generic and specific enough 😄  it will probably make more sense if I add a comment.